File: gpu/GPUTutorial/transpose.c
/*--------------------------------------------------------------------*/
/* C Library GPU Tutorial            */
/* written by Viktor K. Decyk, UCLA  */

#include "transpose.h"

void transpose0(float a[], float b[], int nx, int ny) {
/* simple 2d transpose of length nx, ny */
/* a = transpose(b)                     */
/* local data */
   int j, k;

   for (k = 0; k < ny; k++) {
      for (j = 0; j < nx; j++) {
         a[k+ny*j] = b[j+nx*k];
      }
   }

   return;
}

void  transpose2(float a[], float b[], int mx, int my, int nx, int ny) {
/* segmented 2d transpose of length nx, ny, with block size mx, my */
/* a = transpose(b)                                                */
/* local data */
   int j, k, idx, idy, joff, koff, nbx, nby, mxv, mxm, mym;
/* scratch fast, local array */
   float s[(mx+1)*my];
/* nbx/nby = number of blocks in x/y */
   nbx = (nx - 1)/mx + 1; nby = (ny - 1)/my + 1;
   mxv = mx + 1;

   for (idy = 0; idy < nby; idy++) {
      koff = my*idy;
      mym = my < ny-koff ? my : ny-koff;
      for (idx = 0; idx < nbx; idx++) {
         joff = mx*idx;
         mxm = mx < nx-joff ? mx : nx-joff;

/* copy in from slow memory with stride 1 into block of fast memory */
         for (k = 0; k < mym; k++) {
            for (j = 0; j < mxm; j++) {
               s[j+mxv*k] = b[j+joff+nx*(k+koff)];
            }
         }

/* copy out to slow memory with stride 1 from block of fast memory */
         for (j = 0; j < mxm; j++) {
            for (k = 0; k < mym; k++) {
               a[k+koff+ny*(j+joff)] = s[j+mxv*k];
            }
         }

      }
   }

   return;
}

/* Interfaces to Fortran */

void ctranspose0_(float *a, float *b, int *nx, int *ny) {
   transpose0(a,b,*nx,*ny);
   return;
}

void ctranspose2_(float *a, float *b, int *mx, int *my, int *nx,
                  int *ny) {
   transpose2(a,b,*mx,*my,*nx,*ny);
   return;
}

File: gpu/GPUTutorial/transpose.f90
!-----------------------------------------------------------------------
! Fortran Library GPU Tutorial
! written by Viktor K. Decyk, UCLA
      module transpose
      implicit none
!
      contains
!
      subroutine transpose0(a,b)
! simple 2d transpose of length nx, ny
! a = transpose(b)
      implicit none
      real, dimension(:,:) :: a, b
! local data
      integer :: j, k, nx, ny
! nx/ny = size of arrays in x/y
      nx = min(size(a,2),size(b,1)); ny = min(size(a,1),size(b,2))
!
      do k = 1, ny
         do j = 1, nx
            a(k,j) = b(j,k)
         enddo
      enddo
!
      end subroutine
!
      subroutine transpose2(a,b,mx,my)
! segmented 2d transpose of length nx, ny, with block size mx, my
! a = transpose(b)
      implicit none
      integer :: mx, my
      real, dimension(:,:) :: a, b
! local data
      integer :: j, k, nx, ny, idx, idy, joff, koff, nbx, nby
! scratch fast, local array
      real, dimension(mx+1,my) :: s
! nx/ny = size of arrays in x/y
      nx = min(size(a,2),size(b,1)); ny = min(size(a,1),size(b,2))
! nbx/nby = number of blocks in x/y
      nbx = (nx - 1)/mx + 1; nby = (ny - 1)/my + 1
!
      do idy = 0, nby-1
         koff = my*idy
         do idx = 0, nbx-1
            joff = mx*idx
!
! copy in from slow memory with stride 1 into block of fast memory
            do k = 1, min(my,ny-koff)
               do j = 1, min(mx,nx-joff)
                  s(j,k) = b(j+joff,k+koff)
               enddo
            enddo
!
! copy out to slow memory with stride 1 from block of fast memory
            do j = 1, min(mx,nx-joff)
               do k = 1, min(my,ny-koff)
                  a(k+koff,j+joff) = s(j,k)
               enddo
            enddo
!
         enddo
      enddo
!
      end subroutine
!
      end module

